finality-aleph/src/sync/mock/backend.rs

```diff
@@ -95,41 +81,40 @@ impl Backend {
             .expect("notification receiver is open");
     }
 
-    pub fn import(&self, header: MockHeader) {
+    fn prune(&self) {
         let mut storage = self.inner.lock();
-
-        let parent_id = match header.parent_id() {
-            Some(id) => id,
-            None => panic!("importing block without a parent: {header:?}"),
-        };
-
-        if storage.blockchain.contains_key(&header.id()) {
-            panic!("importing an already imported block: {header:?}")
-        }
-
-        if !storage.blockchain.contains_key(&parent_id) {
-            panic!("importing block without an imported parent: {header:?}")
-        }
-
-        if header.id().number() != parent_id.number() + 1 {
-            panic!("importing block without a correct parent: {header:?}")
+        let mut mark_prune = storage.finalized.clone();
+        let mut mark_keep = HashSet::new();
+        mark_keep.insert(mark_prune.pop().expect("should contain at least genesis"));
+        let mut mark_prune = HashSet::<_>::from_iter(mark_prune.into_iter());
+        let mut prune = HashSet::new();
+        let ids = storage.blockchain.keys().clone();
+        let mut chunk = HashSet::new();
+        for id in ids {
+            let mut id = id.clone();
+            loop {
+                if mark_prune.contains(&id) {
+                    prune.extend(chunk.clone().drain());
+                    mark_prune.extend(chunk.drain());
+                    break;
+                }
+                if mark_keep.contains(&id) {
+                    mark_keep.extend(chunk.drain());
+                    break;
+                }
+                chunk.insert(id.clone());
+                id = storage
+                    .blockchain
+                    .get(&id)
+                    .expect("should be there")
+                    .header()
+                    .parent_id()
+                    .expect("genesis should be marked");
+            }
```

Comment:
I think this is overcomplicated. The easy way to do this is:
```rust
        let last_finalized = self.top_finalized()?;
        let mut storage = self.inner.lock();

        let finalized = &storage.finalized;
        let mut to_prune = vec![];
        for id in storage.blockchain.keys() {
            if id.number <= last_finalized.header.id.number && !finalized.contains(id) {
                to_prune.push(id)
            }
        }
```

Reply:
this can leave some partially pruned branches hanging in the air (literally...)

Reply:
The hanging branches will be cleared later

Reply:
Ok, the question is how does the Substrate db behave in such situation. I was under the impression that the forks should become inaccessible immediately.

Reply:
hmm then we can do smth like this:
```rust
        let last_finalized = self.top_finalized()?;
        let mut storage = self.inner.lock();

        let finalized = &storage.finalized;
        let mut to_prune = vec![];
        for id in storage.blockchain.keys() {
            if id.number <= last_finalized.header.id.number && !finalized.contains(id) {
                to_prune.push(id);
            } else if  id.number > last_finalized.header.id.number && !is_predecessor(id, last_finalized) {
               to_prune.push(id)
           }
        }
```
its not the most efficient way to do it but its more readable ;)

Reply:
fixed, but you'll pay the electricity bill...

Reply:
I think even this can be simplified to:
```rust
if !finalized.contains(id) && !is_predecessor(id, last_finalized) {
    to_prune.push(id)
}
```
right?

Reply:
yup, fixed
I've even used `filter`!